Plutarch/DataRepr/Internal.hs

```diff
@@ -627,3 +646,122 @@ class IsBuiltinList a where
 
 instance IsBuiltinList (PDataRecord l) where
   dataListFrom = punsafeCoerce
+
+----------------------- HRecP and friends -----------------------------------------------
+
+type HRecPApply :: [(Symbol, PType)] -> S -> [(Symbol, Type)]
+type family HRecPApply as s where
+  HRecPApply ('(name, ty) ': rest) s = '(name, Reduce (ty s)) ': HRecPApply rest s
+  HRecPApply '[] s = '[]
+
+newtype HRecP (as :: [(Symbol, PType)]) (s :: S) = HRecP (HRecGeneric (HRecPApply as s))
+
+instance Reducible (HRecP as s) where type Reduce (HRecP as s) = HRecGeneric (HRecPApply as s)
+
+newtype Flip f a b = Flip (f b a)
```

Comment:
We have these helper types in `.Builtin` as well right?

Reply:
Yeah, they're quite small, and more
similar to a lambda expression than a function definition. Not sure whether it's the right choice to share them.